### PR TITLE
re-enable the 0-arg MersenneTwister() constructor

### PR DESCRIPTION
**Pull request by rfourquet**
Cf. #16984 for context, which deprecated `MersenneTwister() = MersenneTwister(0)`. 
The next step here is to re-enable this constructor to be equivalent to `srand(MersenneTwister(0))`. 
I think that it's good for predictability that for an RNG type `T`, `T([seed])` produces an RNG object equivalent to `srand(T(), [seed]))`.
Accordingly, `srand(RandomDevice())` is also enabled (as a no-op).
I put WIP as I'm not satisfied with my wording in the docstrings.

### CODE REVIEW

**Comment by rfourquet** [comment]
Rebased. I will merge in a few days if no objection.

**Review by StefanKarpinski — APPROVED** [review]

File: base/random/RNGs.jl [hunk 1]
```diff
@@ -40,9 +40,13 @@ end # os-test
 
 Create a `RandomDevice` RNG object.
 Two such objects will always generate different streams of random numbers.
+The entropy is obtained from the operating system.
 """
 RandomDevice
 
+RandomDevice(::Void) = RandomDevice()
```
Comment (tkelman) [inline base/random/RNGs.jl]: Why is this necessary?

Comment (rfourquet) [reply base/random/RNGs.jl]: It's not. For `MersenneTwister`, I introduced it for convenience (but still not necessary), and I thought it could be a nice API (not yet documented): when you want to initialize your RNG, you can default your seed to `nothing` to have the behavior of non-reproducibility. WIthout `nothing`, you have to handle the 2 cases:
```
seed = Nullable{Int}()
...
rng = isnull(seed) ? MersenneTwister() : MersenneTwister(seed)
# vs:
rng = MersenneTwister(get(seed, nothing))
```

Comment (tkelman) [reply base/random/RNGs.jl]: couldn't you splat an empty tuple? we don't often use nothing as an input flag, other than unset keywords

Comment (rfourquet) [reply base/random/RNGs.jl]: > couldn't you splat an empty tuple? we don't often use nothing as an input flag, other than unset keywords

I guess splatting a tuple would work... but it seams more natural to me to use `nothing`, and your mention of keyword is helpful: it's easy to have an API like `do_computation(...; seed=nothing)` and feed the RNG directly with `seed` without having to splat a tuple. I remember few times having found inconvenient to not be able to give a default value to a seed which could mean "seed randomly". A couple of relevant (kind of) examples which receive a seed as parameter which could benefit from allowing `nothing`: https://github.com/denizyuret/Knet.jl/pull/105/files and https://github.com/JuliaGraphs/LightGraphs.jl/pull/579/files
